File: 58.翻转单词序列/58.翻转单词顺序列.cpp
class Solution {
public:
    string ReverseSentence(string str) {
        auto size = str.size();
        if (size==0) return "";
        int mark = 0;
        str += ' ';
        for (int i = 0;i<size+1;i++){
            if (str[i] == ' '){
                ReverseWord(str,mark,i-1);
                mark = i + 1;
            }
        }
        str = str.substr(0,size);
        ReverseWord(str,0,size-1);
        return str;
    }
        void ReverseWord (string &str, int l, int r){
        while(l < r){
            swap(str[l], str[r]);
            ++l;
            --r;
        }
    }
};

File: 58.翻转单词序列/58.翻转单词顺序列.py
# -*- coding:utf-8 -*-
class Solution:
    def ReverseSentence(self, s):
        # write code here
        i = 0
        while(i<len(s) and s[i]==" "):
                i += 1
        if i == len(s):
            return s
        stack = []
        ans = ""
        i = 0
        while(i<len(s)):
            while(i<len(s) and s[i]==" "):
                i += 1
            temp = ""
            while(i < len(s) and s[i]!=" "):
                temp += s[i]
                i += 1
            stack.append(temp)
        ans += stack.pop()
        while(len(stack)>0):
            ans += " "
            ans += stack.pop()
        return ans
           